File: Java/trapping-rain-water.java
/**
 * Problem: https://leetcode.com/problems/trapping-rain-water/
 * 
 * To find the trapped water at position i, we need to find the maximum value of the left elements of i and right elements of i. 
 * Assuming they are maxLeft[i] and maxRight[i]. 
 * The trapped water is min(maxLeft[i], maxRight[i]) – height[i] (if this value is larger than 0).
 * 
 * Ref: http://www.lifeincode.net/programming/leetcode-trapping-rain-water-java/
 */
public class Solution {
    public int trap(int[] height) {
        int len = height.length;
        int[] maxLeft = new int[len];
        int[] maxRight = new int[len];
        
        for(int i=1; i<len; i++) {
            maxLeft[i] = Math.max(maxLeft[i-1], height[i-1]);
        }
        for(int i=len-2; i>=0; i--) {
            maxRight[i] = Math.max(maxRight[i+1], height[i+1]);
        }
        
        int total = 0;
        for(int i=0; i<len; i++) {
            total += Math.max(0, Math.min(maxLeft[i], maxRight[i]) - height[i]);
        }
        
        return total;
    }
}

File: Swift/trapping-rain-water.swift
/**
 * Problem Link: https://leetcode.com/problems/trapping-rain-water/
 *
 */

class Solution {
    func trap(_ height: [Int]) -> Int {
        let n = height.count
        if n == 0 {
            return 0
        }
        var maxLeft = [Int]()
        var maxRight = [Int]()
        maxLeft.append(0)
        maxRight.append(0)
        for i in 1..<n {
            maxLeft.append(max(height[i-1], maxLeft.last!))
            maxRight.insert(max(height[n-i], maxRight.first!), at: 0)
        }
        
        var total = 0
        for i in 0..<n {
            total += max(0, min(maxLeft[i], maxRight[i]) - height[i])
        }
        return total
    }
}

class Solution {
    func trap(_ height: [Int]) -> Int {
        var result = 0
        var stack = [Int]()
        for index in 0 ..< height.count {
            while let top = stack.last, height[index] > height[top] {
                stack.removeLast()
                if let last = stack.last {
                    let dist = index - last - 1
                    let bounded_height = min(height[index], height[last]) - height[top]
                    result += dist * bounded_height
                }
            }
            stack.append(index)
        }
        return result
    }
}

class Solution {
    func trap(_ height: [Int]) -> Int {
        var leftMax = 0
        var rightMax = 0
        var left = 0
        var right = height.count - 1
        var result = 0
        while left < right {
            if height[left] < height[right] {
                if height[left] >= leftMax {
                    leftMax = height[left]
                } else {
                    result += leftMax - height[left]
                }
                left += 1
            } else {
                if height[right] >= rightMax {
                    rightMax = height[right]
                } else {
                    result += rightMax - height[right]
                }
                right -= 1
            }
        }
        return result
    }
}
